packages/runner/src/runner.ts

```diff
@@ -475,4 +474,346 @@ export class Runner {
+        hash_type: this._deploymentConfig().sudt_type.hash_type,
+        args: sudtScriptHash,
+      };
+      let sudtChangeCustodian: Cell = {
+        cell_output: {
+          lock: custodianLock,
+          type: custodianType,
+          capacity: "0x0",
+        },
+        data: utils.toBigUInt128LE(
+          BigInt(inputSudtAmountSum) - BigInt(targetSudtAmount)
+        ),
+      };
+      const minimalCapacity = minimalCellCapacity(sudtChangeCustodian);
+      sudtChangeCustodian.cell_output.capacity =
+        "0x" + minimalCapacity.toString(16);
+      outputCkbCapacitySumForSudtCustodianCells += minimalCapacity;
+      txSkeleton = txSkeleton.update("outputs", (outputs) => {
+        return outputs.push(sudtChangeCustodian);
+      });
+    }
+    if (
+      BigInt(inputCkbCapacitySum) ===
+      BigInt(ckbWithdrawalCapacity) +
+        BigInt(outputCkbCapacitySumForSudtCustodianCells)
+    ) {
+      // so lucky with it, just return
+      return txSkeleton;
+    }
+    // build ckb change custodian cell
+    const custodianLock: Script = {
+      code_hash: this._deploymentConfig().custodian_lock.code_hash,
+      hash_type: this._deploymentConfig().custodian_lock.hash_type,
+      args: DEFAULT_CUSTODIAN_LOCK_ARGS,
+    };
+    let ckbChangeCustodian: Cell = {
+      cell_output: {
+        lock: custodianLock,
+        type: undefined,
+        capacity: "0x0",
+      },
+      data: utils.toBigUInt128LE(BigInt(0)),
+    };
+    const minimalCapacity = minimalCellCapacity(ckbChangeCustodian);
+    const changeCapacity =
+      BigInt(inputCkbCapacitySum) -
+      BigInt(ckbWithdrawalCapacity) -
+      BigInt(outputCkbCapacitySumForSudtCustodianCells);
+    if (BigInt(changeCapacity) >= BigInt(minimalCapacity)) {
+      // With enough input ckb capacity to build a minimal change cell
+      ckbChangeCustodian.cell_output.capacity =
+        "0x" + changeCapacity.toString(16);
+      txSkeleton = txSkeleton.update("outputs", (outputs) => {
+        return outputs.push(ckbChangeCustodian);
+      });
+      return txSkeleton;
+    }
+    // Gonna collect some more ckb to build the change cell
+    for (const cell of validCustodianCells) {
+      if (!cell.cell_output.type) {
+        const key = getInputKey(cell);
+        if (previousInputs.has(key)) {
+          continue;
+        }
+        previousInputs.add(key);
+        const inputCkbCapacity = BigInt(cell.cell_output.capacity);
+        inputCkbCapacitySum += inputCkbCapacity;
+        txSkeleton = txSkeleton.update("inputs", (inputs) => {
+          return inputs.push(cell);
+        });
+        if (
+          BigInt(inputCkbCapacitySum) >=
+          BigInt(ckbWithdrawalCapacity) +
+            BigInt(outputCkbCapacitySumForSudtCustodianCells) +
+            BigInt(minimalCapacity)
+        ) {
+          break;
+        }
+      }
+    }
+    // TODO collect ckb from other sudt custodian cells if all non sudt custodian cells' ckb capacity is insufficient,
```

Comment:
I think we can add checkings in godwoken: if someone chooses to withdraw SUDT, he/she must withdraw CKB of enough capacity to store the SUDTs. Also, the left CKBs must be enough to hold a left-over custodian cell.

If above rules are not met, we should reject the withdrawal request. In this case, TODO here won't be needed.

Reply:
updated.